lunabot_control/src/mpc.cpp

```diff
@@ -159,59 +105,75 @@ double MPC::smooth_clamp_(double x, double a, double b) {
   return a + (b - a) * (1 + sin(M_PI * x / 2)) / 2;
 }
 
-std::vector<Eigen::MatrixXd> MPC::normal_distribute_(Eigen::MatrixXd means,
-                                                     Eigen::MatrixXd std_devs, int count) {
-  std::vector<Eigen::MatrixXd> random_vels(count);
+std::pair<std::vector<Eigen::MatrixXd>, std::vector<Eigen::MatrixXd>>
+MPC::normal_distribute_(Eigen::MatrixXd means, Eigen::MatrixXd std_devs, int count) {
+  std::vector<Eigen::MatrixXd> random_vels_raw(count);
+  std::vector<Eigen::MatrixXd> random_vels_clamped(count);
+
+#pragma omp parallel for
   for (int i = 0; i < count; ++i) {
     Eigen::MatrixXd random_vel(means.rows(), means.cols());
+    Eigen::MatrixXd random_vel_clamped(means.rows(), means.cols());
 
+#pragma omp parallel for
     for (int j = 0; j < means.rows(); ++j) {
+
+#pragma omp parallel for
       for (int k = 0; k < means.cols(); ++k) {
         std::random_device rd;
         std::default_random_engine generator(rd());
         std::normal_distribution<double> distribution(means(j, k), std_devs(j, k));
         double rand_vel = distribution(generator);
-        /*
+        double rand_vel_clamped = 0;
         if (k == 0) {
-          rand_vel = smooth_clamp_(rand_vel, lin_lim_[0], lin_lim_[1]);
+          rand_vel_clamped = smooth_clamp_(rand_vel, lin_lim_[0], lin_lim_[1]);
         } else {
-          rand_vel = smooth_clamp_(rand_vel, ang_lim_[0], ang_lim_[1]);
+          rand_vel_clamped = smooth_clamp_(rand_vel, ang_lim_[0], ang_lim_[1]);
         }
-        */
         random_vel(j, k) = rand_vel;
+        random_vel_clamped(j, k) = rand_vel_clamped;
       }
     }
-    random_vels[i] = random_vel;
+    random_vels_raw[i] = random_vel;
+    random_vels_clamped[i] = random_vel_clamped;
   }
-  return random_vels;
+  return std::make_pair(random_vels_clamped, random_vels_raw);
 }
 
 Eigen::MatrixXd MPC::mean_(std::vector<std::pair<Eigen::MatrixXd, double>> rollouts) {
   Eigen::MatrixXd mean = Eigen::MatrixXd::Zero(this->horizon_length_, 2);
+#pragma omp parallel for
   for (int i = 0; i < this->top_rollouts_; ++i) {
+#pragma omp parallel for
     for (int j = 0; j < rollouts[i].first.rows(); ++j) {
-      mean(j, 0) += rollouts[i].first(j, 3);
-      mean(j, 1) += rollouts[i].first(j, 4);
+      mean(j, 0) += rollouts[i].first(j, 5);
```

Comment:
This is also on me but maybe have comments / better way to understand this than just idx 5 / 6

Reply:
Yeah, lets make some #define statements for all these indexes?